### PR TITLE
Show controls as read only based on tenant permissions

### PR DESCRIPTION
**Pull request by jakubp-eliatra**
### Description
Adds a proper read only mode support for read only tenants.

A new capability `hide_for_read_only` as array is utilized to recognize the capabilities which should be disable in read only mode. If a read only tenant is recognized the capabilities associated with writing access such as 'createNew' and 'saveQuery' are then set to false.

### Category.
New feature

### Why these changes are required?
- Resolves https://github.com/opensearch-project/security/issues/2701.

### What is the old behavior before changes and new behavior after changes?
For example re-arranging visualisations is not possible in such mode now.

### CODE REVIEW

**Comment by davidlago** [comment]
@jakubp-eliatra , the DCO check is failing on this PR. I'm going to hold off on approving the CI run since we'll have to re-run anyway.

**Comment by jakubp-eliatra** [comment]
Just updated the PR. Two remarks:

1. Whenever the capabilities route is called, we perform a request to authinfo in order to get information about the user's tenant. This request requires an authorization header.
We did run into the problem that we don't have the authorization header set on the request. The reason for this is that the capabilities route is ignored by the auth handler, which means that the auth header is never added [here.](https://github.com/opensearch-project/security-dashboards-plugin/blob/ab96df70226c4d5155c301a5123b569df56eba65/server/auth/types/authentication_type.ts#L77)
We solved this by always passing the auth type to the capabilities switcher logic, and using the auth type and the cookie to build the authorization header [here.](https://github.com/opensearch-project/security-dashboards-plugin/blob/ab96df70226c4d5155c301a5123b569df56eba65/server/plugin.ts#L169)

Another possible solution, and maybe something to address either way in the future, would be to stop ignoring the capabilities route.
That route is defined as auth being optional, and that case doesn't seem to be handled by the authentication logic right now. To handle that case, we could check the route's auth settings before returning a 401 Unauthorized. For example like this:

```
if (request.route.options?.authRequired === 'optional') {  
  return toolkit.notHandled(); // Maybe toolkit.authorized()?
}
```

That way, the request will not throw an error if it is not authenticated, but if it is authenticated, the regular process would continue.

2. As discussed, this PR assumes that a complementary PR for Dashboards would get created that adds `hide_for_read_only` property to capabilities' providers like so:
`export const capabilitiesProvider = () => ({   dashboard: {     createNew: true,     show: true,     showWriteControls: true,     saveQuery: true,     hide_for_read_only: ['createNew', 'saveQuery'],   }, });`

The new property is an array that contains the names of capabilities that will get toggled to false (for read only tenants) using the logic from this PR.

One idea that also came up was to make sure that all apps register capability 'showWriteControls', just as the Dashboards app does. That one capability would fit well for the read only use case, if it takes precedence over other capabilities that are used for write controls. But since we already had a discussed solution, we went that path.

**Comment by jakubp-eliatra** [comment]
We took our time to estimate the efforts of adjusting internal and external plugins to read only mode. Here are the results:

**a) internal plugins that need modification:**

1. advanced_settings* (save)
2. dashboard* (createNew, showWriteControls)
3. data*
4. discover* (save)
5. home*
6. management*
7. opensearch_dashboards_legacy (showWriteControls)
8. opensearch_dashboards_react (advancedSettings, navLinks)
9. saved_objects_management* (delete)
10. timeline* (save)
11. vis_builder
12. visualize* (show)
13. console* + dev_tools (!) (show, save)

_**asterisk (*)** - registers capabilities but doesn't do anything with them OR uses capabilities in a way unrelated with logic that should be updated for read only mode; they might need little adjustments or not at all, **but they should at least be tested**
**(x, y)** - known capabilities used within the plugin_

**b) external plugins that need modification:**

- dashboards-maps - https://github.com/opensearch-project/dashboards-maps/blob/main/server/saved_objects/map_saved_object.ts
- dashboards-observability - https://github.com/opensearch-project/dashboards-observability/blob/main/server/saved_objects/observability_saved_object.ts

Those are the only two out of external plugins that we found they have something to do with capabilities at all. There doesn't seem to be any logic related with read only mode inside of those plugins. By that I mean that I checked the codebase in general, wandering from file to file and searched for buzzwords such as capabilities + readonly/disabled (obviously with excluding cases where it's just UI logic, such as disabling a form field somewhere).

**Due to the fact that for example maps plugin contains customizable layers of maps (and user can set zoom etc.), I believe it would be best for these two to incorporate some logic related with read only mode for those two plugins.** In other words, a support for read only mode would make for a better UX in case of those plugins.

**c) my opinion on the research that's been done:**

Luckily, most of changes are limited to main Dashboards repo, however in the future more and more plugins might need to be aware of how the read only mode works and what property should be registered (hide_for_read_only array if that's what we're going with) to make it work properly.

**Therefore merging of the read only PR needs go along with a proper documentation.**

Summing up, it seems that there is no ideal solution. We either:
1. have a list of capabilities are related with read only logic to toggle them (like showWriteControls, show etc.) - heuristic solution but there's the guessing factor (we eliminated this approach here: https://github.com/opensearch-project/security/issues/2701) 
2. we have them explicitly listed within the hide_for_read_only_property and we adjust some logic within internal/external plugins described above.

**Comment by cwperks** [comment]
@jakubp-eliatra Are there any tests that can be written for this change? Unit, integration of functional tests?

**Review by scrawfor99 — COMMENTED** [review]
Hi @jakubp-eliatra, thank you for putting this together. Overall, everything makes sense to me. That being said, we will want to add some tests to verify the functionality is expected. Since we are making a pretty significant UI change when the tenant is read-only, you should be able to use Cypress to check that an element which is present with not read-only is no longer present for a read-only tenant.

File: server/plugin.ts [hunk 1]
```diff
@@ -86,6 +87,107 @@ export class SecurityPlugin implements Plugin<SecurityPluginSetup, SecurityPlugi
...
+    if (!request.headers || !request.headers.referer) {
+      return false;
+    }
+
+    try {
+      const url = new URL(request.headers.referer as string);
+      const pathsToIgnore = ['login', 'logout', 'customerror'];
+      return pathsToIgnore.includes(url.pathname?.split('/').pop() || '');
+    } catch (error: any) {
+      this.logger.error(`Could not parse the referer for the 1: ${error.stack}`);
```
Comment (scrawfor99) [inline server/plugin.ts:100]: Nit: I am a bit confused by the message. Is this phrasing from somewhere else? I am not sure what 'the 1' is referring to here.

**Review by scrawfor99 — COMMENTED** [review]
Looks good, just one comment about an error message

File: server/plugin.ts (hunk 1 above)
Comment (kajetan-nobel) [reply server/plugin.ts]: Good catch 😅 It had to be an accidental typo.

**Comment by peternied** [comment]
@jakubp-eliatra We've got an upcoming 2.10 release and I'm reviewing outstanding pull requests, would you like this pull request to be considered for inclusion?

It looks like this PR has stalled for a little, but is there anything that can be done to help?

**Comment by kajetan-nobel** [comment]
Hi @peternied,  I've taken this task over from @jakubp-eliatra. 

It has to wait due to the probably different approach that we'll take (context: https://github.com/opensearch-project/security/issues/2701#issuecomment-1661238661)

**Comment by peternied** [comment]
@kajetan-nobel I'm going to change this pull request to draft while there is still back-and-forth around the design.  Please let the reviews know its ready or close out if that makes more sense.

**Comment by kajetan-nobel** [comment]
@peternied please remove it from draft state. I completely forgot to ask it recently 🥹

Also, could a maintainers of this repository please review this PR for me please? Thank you.
For context it's a supplement for https://github.com/opensearch-project/OpenSearch-Dashboards/pull/4498

File: server/plugin.ts [hunk 2]
```diff
@@ -240,12 +135,10 @@ export class SecurityPlugin implements Plugin<SecurityPluginSetup, SecurityPlugi
     // Register server side APIs
     defineRoutes(router);
     defineAuthTypeRoutes(router, config);
+
     // set up multi-tenent routes
-    if (config.multitenancy?.enabled) {
+    if (config.multitenancy.enabled) {
```
Comment (DarshitChanpura) [inline server/plugin.ts:140]: Why remove the ternary check here?

File: server/readonly/readonly_service.ts [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+  }
+
+  async hideForReadonly(
+    request: OpenSearchDashboardsRequest,
+    capabilites: Capabilities,
+    hideCapabilities: Partial<Capabilities>
+  ): Promise<Partial<Capabilities>> {
+    return (await this.isReadonly(request)) ? merge(capabilites, hideCapabilities) : capabilites;
+  }
+}
```
Comment (DarshitChanpura) [inline server/readonly/readonly_service.ts:104]: Could we add tests to verify this change?

File: server/plugin.ts [hunk 4]
```diff
@@ -87,107 +83,6 @@ export class SecurityPlugin implements Plugin<SecurityPluginSetup, SecurityPlugi
...
-  registerSwitcher(
-    core: CoreSetup,
-    securityClient: SecurityClient,
-    auth: IAuthenticationType,
-    securitySessionStorageFactory: SessionStorageFactory<SecuritySessionCookie>
-  ) {
-    core.capabilities.registerSwitcher(
-      this.capabilitiesSwitcher(securityClient, auth, securitySessionStorageFactory)
-    );
-  }
```
Comment (DarshitChanpura) [inline server/plugin.ts]: For my understanding, what is the difference between registering a switcher vs the new way?

File: server/readonly/readonly_service.ts (hunk 3 above)
Comment (kajetan-nobel) [reply server/readonly/readonly_service.ts:104]: Sure, gonna do it

File: server/plugin.ts (hunk 2 above)
Comment (kajetan-nobel) [reply server/plugin.ts:140]: probably missed that during development or merge, rebase

File: server/plugin.ts (hunk 4 above)
Comment (kajetan-nobel) [reply server/plugin.ts]: @DarshitChanpura the `registerSwitcher` way introduced a `hide_for_read_only` which was extending capabilities type and was a messy way of saying "change it for me". With a new approach, we can use the existing `registerSwitcher` and separate logic to check if the current tenant is read-only without bruising capabilities definition. We're then even more able to define behaviors of each plugin in case of read-only.

**Comment by kajetan-nobel** [comment]
FYI: working on additional tests

**Comment by DarshitChanpura** [comment]
@kajetan-nobel I've marked this PR as draft as it is still being worked on and is pending some merges in core. Feel free to mark this as ready for review once all changes are complete.

**Comment by kajetan-nobel** [comment]
@DarshitChanpura @peternied I've added tests for `readonly_service`. It's again ready for review.

**Comment by kajetan-nobel** [comment]
The tests here will fail as long as https://github.com/opensearch-project/OpenSearch-Dashboards/pull/4498 will be not merged first.

**Comment by kajetan-nobel** [comment]
@peternied @davidlago @cliu123 @cwperks @RyanL1997 
I need your help with the review here :)

File: server/readonly/readonly_service.ts [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+
+      const authInfo = await this.securityClient.authinfo(request, headers);
+
+      if (!authInfo.user_requested_tenant && cookie) {
+        authInfo.user_requested_tenant = cookie.tenant;
+      }
+
+      return authInfo && this.isReadOnlyTenant(authInfo);
+    } catch (error: any) {
+      return false;
```
Comment (peternied) [inline server/readonly/readonly_service.ts:102]: Lets log this error

File: server/readonly/readonly_service.ts [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+      return pathsToIgnore.includes(url.pathname?.split('/').pop() || '');
+    } catch (error: any) {
+      this.logger.error(`Could not parse the referer for the capabilites: ${error.stack}`);
+    }
+  }
+
+  isReadOnlyTenant(authInfo: OpenSearchAuthInfo): boolean {
+    const currentTenant = authInfo.user_requested_tenant || globalTenantName;
+
+    // private tenant is not affected
```
Comment (peternied) [inline server/readonly/readonly_service.ts:68]:
```suggestion
    // private tenants are isolated to individual users that always have read/write permissions
```

File: server/readonly/readonly_service.ts [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+
+  isAnonymousPage(request: OpenSearchDashboardsRequest) {
+    if (!request.headers || !request.headers.referer) {
+      return false;
+    }
+
+    try {
+      const url = new URL(request.headers.referer as string);
+      const pathsToIgnore = ['login', 'logout', 'customerror'];
+      return pathsToIgnore.includes(url.pathname?.split('/').pop() || '');
```
Comment (peternied) [inline server/readonly/readonly_service.ts:59]: [Refactor] These values seem important, is there another place where they are listed?  Doesn't seem like it should be in this file since it would also be part of determining if redirection to login occur during request routing.

File: server/readonly/readonly_service.ts [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+    } catch (error: any) {
+      this.logger.error(`Could not parse the referer for the capabilites: ${error.stack}`);
```
Comment (peternied) [inline server/readonly/readonly_service.ts:61]: If there is an error, shouldn't it be throw?

That error message doesn't seem accurate based on the code above, no capabilities are listed?

File: server/readonly/readonly_service.ts [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+
+  isReadOnlyTenant(authInfo: OpenSearchAuthInfo): boolean {
+    const currentTenant = authInfo.user_requested_tenant || globalTenantName;
+
+    // private tenant is not affected
+    if (isPrivateTenant(currentTenant)) {
+      return false;
+    }
+
+    return authInfo.tenants[currentTenant] !== true;
```
Comment (peternied) [inline server/readonly/readonly_service.ts:73]: Lets use a variable to make it clear what is being extracted from the auth info data structure
```suggestion
    let readWriteAccess = authInfo.tenants[currentTenant];
    return !readWriteAccess;
```

File: server/readonly/readonly_service.ts [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+
+      if (!this.auth.requestIncludesAuthInfo(request) && cookie) {
+        headers = this.auth.buildAuthHeaderFromCookie(cookie, request);
+      }
+
+      const authInfo = await this.securityClient.authinfo(request, headers);
+
+      if (!authInfo.user_requested_tenant && cookie) {
+        authInfo.user_requested_tenant = cookie.tenant;
+      }
```
Comment (peternied) [inline server/readonly/readonly_service.ts:98]: I'm not fully familiar with these systems, could this be inside of the `authinfo(...)`, why have it here?

**Review by peternied — CHANGES_REQUESTED** [review]
I've left some comments around the code - some of those questions might be naive as I don't often work in this space.  Those questions and callouts assign the general shape of this change looks good, thanks for getting the base feature built up across Dashboards.

File: server/readonly/readonly_service.ts (hunk 10 above)
Comment (kajetan-nobel) [reply server/readonly/readonly_service.ts:98]: @peternied I've experienced (at least at my setup) that `authInfo` (at this point) doesn't contain information about `requested_tenant`. Usually, it should be within `authInfo` but it looks like there are some cases where it's not.

**Comment by peternied** [comment]
@jakubp-eliatra Any overall updates? I only see a few couple of changes.

File: server/readonly/readonly_service.ts (hunk 5 above)
Comment (kajetan-nobel) [reply server/readonly/readonly_service.ts:99]: Fixed

**Comment by kajetan-nobel** [comment]
> @jakubp-eliatra Any overall updates? I only see a few couple of changes.

@peternied I'm waiting for your response to my comments here. I don't know why some changes aren't shown here but are in the files...

**Comment by peternied** [comment]
Thanks @kajetan-nobel - Can you look into the CI failures?  They might be related to these changes, or do we need to wait on the Dashboards codebase to accept the change before CI will pass - if so that OK - I'll approve when we've got clean CI in any case.

File: server/readonly/readonly_service.ts (hunk 7 above)
Comment (peternied) [reply server/readonly/readonly_service.ts]: @kajetan-nobel Did you want to follow up on this change, if not could you see around creating an issue to track?

**Comment by kajetan-nobel** [comment]
@peternied PR extends things from dashboards so until https://github.com/opensearch-project/OpenSearch-Dashboards/pull/4498 is merged it will fail.

File: server/readonly/readonly_service.ts (hunk 7 above)
Comment (kajetan-nobel) [reply server/readonly/readonly_service.ts]: @peternied this one still occurs as unresolved 😢

**Comment by peternied** [comment]
@kajetan-nobel Looks like the downstream change was merged - I've hit the 'rebase' button and will be looking in to the results of the CI.

File: server/readonly/readonly_service.ts [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+    if (!request.headers || !request.headers.referer) {
+      return false;
+    }
+
+    const url = new URL(request.headers.referer as string);
+    return ANONYMOUS_ROUTES.some((path) => url.pathname?.includes(path));
+  }
+
+  isReadOnlyTenant(authInfo: OpenSearchAuthInfo): boolean {
+    const currentTenant = authInfo.user_requested_tenant || globalTenantName;
```
Comment (DarshitChanpura) [inline server/readonly/readonly_service.ts:61]: will this fail without multi-tenancy enabled checK? similar to one below

File: common/index.ts [hunk 12]
```diff
@@ -38,6 +38,19 @@ export const OPENID_AUTH_LOGOUT = '/auth/openid/logout';
...
+  LOGIN_PAGE_URI,
+  CUSTOM_ERROR_PAGE_URI,
+  API_AUTH_LOGIN,
+  API_AUTH_LOGOUT,
+  OPENID_AUTH_LOGIN,
+  ANONYMOUS_AUTH_LOGIN,
+  OPENID_AUTH_LOGOUT,
+  SAML_AUTH_LOGOUT,
+  ANONYMOUS_AUTH_LOGOUT,
+  SAML_AUTH_LOGIN,
```
Comment (DarshitChanpura) [inline common/index.ts:51]: is this a verified exhaustive list of anonymous routes?

**Review by peternied — DISMISSED** [review]

File: common/index.ts (hunk 12 above)
Comment (kajetan-nobel) [reply common/index.ts:51]: @DarshitChanpura I'm on it verifying

File: server/readonly/readonly_service.ts (hunk 11 above)
Comment (kajetan-nobel) [reply server/readonly/readonly_service.ts:61]: gonna verify it

File: common/index.ts (hunk 12 above)
Comment (kajetan-nobel) [reply common/index.ts:51]: @DarshitChanpura FYI: I was hinted that maybe there is a possibility to check from the route itself, and this list would be a backup. I'm on it.

**Review by scrawfor99 — APPROVED** [review]
This looks good to me; should be all set once DC's comments are addressed.

File: server/readonly/readonly_service.ts (hunk 11 above)
Comment (kajetan-nobel) [reply server/readonly/readonly_service.ts:61]: @DarshitChanpura the function is called after we've checked and disabled multi-tenancy is treated as `__global__` tenant.

File: common/index.ts [hunk 13]
```diff
@@ -38,6 +38,19 @@ export const OPENID_AUTH_LOGOUT = '/auth/openid/logout';
 export const SAML_AUTH_LOGOUT = '/auth/saml/logout';
 export const ANONYMOUS_AUTH_LOGOUT = '/auth/anonymous/logout';
 
+export const ANONYMOUS_ROUTES = [
```
Comment (cwperks) [inline common/index.ts:41]: Can these endpoints be configured in such a way not to call `isReadOnly` at all? I'd prefer not to maintain a list and instead configure this where the route is declared.

File: server/readonly/readonly_service.ts [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+    return !readWriteAccess;
+  }
+
+  async isReadonly(request: OpenSearchDashboardsRequest): Promise<boolean> {
+    // omit for anonymous pages to avoid authentication errors
+    if (this.isAnonymousPage(request)) {
+      return false;
+    }
+
+    if (!this.config?.multitenancy.enabled) {
```
Comment (cwperks) [inline server/readonly/readonly_service.ts:78]: Does this need to take into account the dynamic value of multi-tenancy enabled?

There are multiple places where multitenancy can be disabled. Directly in `opensearch_dashboards.yml`, but also through dynamic config (`config/opensearch-security/config.yml`) as well. To get the dynamic value, many components call the dashboardsinfo endpoint: https://github.com/opensearch-project/security-dashboards-plugin/blob/main/public/apps/account/account-nav-button.tsx#L75-L77

Comment (kajetan-nobel) [reply server/readonly/readonly_service.ts:78]: @cwperks gonna look into that and I'll let you know! Thanks!

Comment (kajetan-nobel) [reply server/readonly/readonly_service.ts:78]: @cwperks gonna look into that and I'll let you know! Thanks!

**Comment by kajetan-nobel** [comment]
@DarshitChanpura @cwperks Regarding unauthorized routes, I found that we've got a similar implementation here https://github.com/opensearch-project/security-dashboards-plugin/blob/31c5a723cb54d8dcb6ab06de4a7a65b1036590e9/server/auth/types/authentication_type.ts#L218-L233
which already contains not exhaustive list. Maybe it would be good to refactor there and reuse here. 

There is also a possibility to use `request.route.options.authRequired` to check if the requested route requires to be auth (lots of contains value `optional`)

**Comment by DarshitChanpura** [comment]
I agree. We should refactor to re-use. 

> There is also a possibility to use request.route.options.authRequired to check if the requested route requires to be auth (lots of contains value optional)

i'm not sure about this one as this seems a change in design or the way API works. This could be done in a follow-up PR

**Comment by cwperks** [comment]
> There is also a possibility to use request.route.options.authRequired to check if the requested route requires to be auth (lots of contains value optional)

This would be my preferred approaches. I think it would be better for maintenance to have an annotation/option where the route is declared that can signal whether the route is protected (auth required) or not. Most instances of routes not requiring auth are going to be login endpoints.

**Comment by kajetan-nobel** [comment]
@DarshitChanpura @cwperks I tried to base it on `authNotRequired`, but it's on a completely different use level, and it would make it much more troublesome. To explain it more:
- `authNotRequired` entirely wants to ignore the `capabilities` route (especially for the login page) while most of isReadonly are calls to `capabilities` 
- `authNotRequired` is based on the current pathname, not the referrer which is kinda important in case of again - capabilities

I'm going rather to keep the current implementation and use there `request.route.options.authRequired`

**Comment by kajetan-nobel** [comment]
@cwperks Regarding https://github.com/opensearch-project/security-dashboards-plugin/pull/1472#discussion_r1392825862 I've added a fix to it.

File: server/readonly/readonly_service.ts (hunk 14 above)
Comment (kajetan-nobel) [reply server/readonly/readonly_service.ts]: Added fix. It's checked through `securityClient`

File: server/auth/types/authentication_type.ts [hunk 15]
```diff
@@ -37,6 +37,11 @@ export interface IAuthenticationType {
   type: string;
   authHandler: AuthenticationHandler;
   init: () => Promise<void>;
+  requestIncludesAuthInfo(request: OpenSearchDashboardsRequest): boolean;
```
[the remark this answers is not in the document]
Comment (kajetan-nobel) [reply server/auth/types/authentication_type.ts:40]: @cwperks Yes, it has to be added because I'm referring to an Interface, not a class in `readonly_service`

[the remark this answers is not in the document]
Comment (cwperks) [reply server/auth/types/authentication_type.ts:40]: Are the abstract methods redundant then? Could the `protected abstract` declarations be removed on the `AuthenticationType` abstract class?

[the remark this answers is not in the document]
Comment (kajetan-nobel) [reply server/auth/types/authentication_type.ts:40]: fixed

File: server/readonly/readonly_service.ts [hunk 16]
```diff
@@ -0,0 +1,119 @@
...
+  ) {
+    super();
+    this.logger = logger;
+    this.securityClient = securityClient;
+    this.auth = auth;
+    this.securitySessionStorageFactory = securitySessionStorageFactory;
+    this.config = config;
+  }
+
+  isAnonymousPage(request: OpenSearchDashboardsRequest) {
```
Comment (cwperks) [inline server/readonly/readonly_service.ts:58]: Is it possible that the readonly service can be applied after a request is authenticated? If can be applied after auth is it possible that this method could not be needed? If this method is required then I'd like to see more test cases around this particular method because its not entirely clear what it means to be an anonymous page (anonymous page === route that doesn't require auth like the login screen?)

If the order of operations is:

1. Request is received
2. auth // if endpoint requires authentication
3. isReadOnly

then it would be guaranteed that `isReadOnly` is only called on authenticated requests.

It looks like `isReadOnly` is passthrough for the routes that do not require authentication so if it could be skipped then this logic may not be necessary.

Comment (kajetan-nobel) [reply server/readonly/readonly_service.ts:58]: I figured someone might want to use the 'isReadonly' method for some custom logic. But it's kinda redundant unless you're planning to use it only within the 'capabilities.registerSwitcher' method.

In simple terms, you won't need it for what it's doing now. But, if you decide to go rogue and use it outside 'capabilities.registerSwitcher', it would be less confusing for a potential use case scenario. 

In summary:

> Is it possible that the readonly service can be applied after a request is authenticated

- for capabilities no
- for custom logic yes

> If it can be applied after auth

not for capabilities

>  its not entirely clear what it means to be an anonymous page (anonymous page === route that doesn't require auth like the login screen?)

exactly what you said - an anonymous page is a page that doesn't require an auth or auth is `optional` (like capabilities route 🫠)

It really depends if we would like to keep the possibility to call it outside `capabilities.registerSwitcher`. If the answer is yes then this method should stay, if not - we can entirely remove it.

**Review by cwperks — APPROVED** [review]

File: common/index.ts (hunk 13 above)
Comment (peternied) [reply common/index.ts]: This was largely addressed with `isAnonymousPage(...)` in a future iteration

File: common/index.ts (hunk 12 above)
Comment (peternied) [reply common/index.ts]: This was resolved by `isAnonymousPage()` function in a more recent commit

File: server/readonly/readonly_service.ts (hunk 7 above)
Comment (peternied) [reply server/readonly/readonly_service.ts]: @kajetan-nobel Can you use the `Resolve conversation` button?  I'm marking this resolved
image

**Review by peternied — APPROVED** [review]
I've reviewed the outstanding comments as well - thanks!
